nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/StringUtils.java

```diff
@@ -78,25 +78,63 @@ public static String join(final Collection collection, String delimiter) {
         return sb.toString().substring(0, sb.lastIndexOf(delimiter));
     }
 
-    public static String padLeft(final String source, int length, char padding) {
-        if (source != null) {
+    public static String padLeft(final String source, int length, String padding) {
```

Comment:
Did you write this yourself, or did you borrow it from somewhere like `commons-lang3`?

Reply:
Hello Mike, This is inspired on both [commons-lang3](https://github.com/apache/commons-lang/blob/d5ffc4dde6236ed5afbedfeb333bb87dceb3fe7a/src/main/java/org/apache/commons/lang3/StringUtils.java#L5247) and the previous implementation, which didn't allow String as padding. 
What do you think about it?

Reply:
As per the null string and invalid input, [here](https://github.com/apache/nifi/pull/3613/files#diff-677bfd660311116a67f3b9356be86acfR58) I made it return the original string if something is bad with the padding or the length, and null if something is bad with the original string, is that ok for you?

Reply:
I think that's the appropriate behavior here in case the underlying libraries for some odd reason have a problem. We don't want people losing the original data altogether.

The reason I asked about the commons-lang3 issue is that if you did derive it from code from commons-lang3 I need to do a license and notice check to make sure that the IP legal blurbs are up to date. It's basically legal, not technical. So when you say inspired is that "you wrote it yourself based on ideas you got" or "I did copy and paste and maybe some tweaks?"

Reply:
The first one :), I knew the copy-paste style is a no-go for obvious legally (and why not moral) reasons, but more importantly I think this way I preserved most of the previous implementation. 

I can post a diff check if you want to take a look at

Reply:
No, legally it's allowed because both are ASLv2 open source code. We'd just need to track attribution.

Reply:
Oh, ok, easier than I thought then. 
Do you think it could be better to pass to a "safer" copy-paste implementation of those methods with an attribution?

Reply:
Yeah. I think it would make more sense for us to do that. Since NiFi and commons-lang3 are ASF projects, I think the L&N will just need to be a javadoc line above each copied method of `StringUtils` that says prominently something like this:

```
/**
 * This method was copied verbatim from Apache commons-lang3. Apache commons-lang3
 * was provided under the ASLv2 and is Copyright 2019 Apache Software Foundation.
```

@joewitt does that seem right to you?

Reply:
If we want to avoid a binary dependency (and I think we do given that we have our own StringUtil in this module for this purpose) then we should do straight up copy-paste of code with attribution.  We have other cases of this.  The proper way to handle this is to update the top level NOTICE file (covers our source) and nifi-assembly/NOTICE (covers our binary distribituion) to indicate what methods we've copied into what module of nifi.

We should not do much of this however.  The intent is not that StringUtils just copies commons-lang3.  We have tons of usage of commons-lang3 binary dependency throughout nifi and we should prefer that approach where it makes sense.

Reply:
Thanks @joewitt . IIUC then you'd go for adding the dependency. I am fine doing it. Shall we go ahead that way @MikeThomsen, do you agree on this ?

Reply:
@adarmiento Joe wants to avoid the dependency for now:

> If we want to avoid a binary dependency (**and I think we do** given that we have our own StringUtil in this module for this purpose [*Emphasis mine* - Mike]

So feel free to just copy the left/right pad methods from `commons-lang3` into our `StringUtils` and add the Javadoc lines I gave you. I'll take a look at the LICENSE and NOTICE files at the root level and give you any updates I think you need to add.

Reply:
Ah ok, I misunderstood the last comment. 
Ok, I'm going ahead with this solution then

Reply:
I also sent you a pull request for the verbiage for the top level `NOTICE` file.

Reply:
ah wow, thank you!

Reply:
I updated the `StringUtils` class with the borrowed Lang methods, I picked the javadoc of those methods as well, do you think I should edit the javadoc removing other Lang stuff references?(for example [the @since annotations right and there](https://github.com/apache/nifi/pull/3613/files#diff-fa10041e6362efbee8ea547b922febccR135)

Reply:
I think we can leave them as-is. We have the top level attribution for the code, so I'm going to merge. Thanks!

Reply:
Thanks to you!